### PR TITLE
Added native browser console app

### PR DESCRIPTION
**Pull request by raphaellondner-okta**
Leverages the default browser (and leveraging any existing Okta browser session) for authentication and OAuth authorization (using Authorization Code Flow w/ PKCE)

### CODE REVIEW

**Comment by raphaellondner-okta** [comment]
@marksztainbok-okta Hope you had a good holiday break! Is there any way you could review and approve this PR this week? Thank you :)

File: Okta.Samples.OpenIDConnect.Console/KeyGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Okta.Samples.OpenIDConnect
+{
+    public class KeyGenerator
+    {
+        public static string GetUniqueKey(int maxSize)
+        {
+            char[] chars = new char[62];
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.Console/KeyGenerator.cs]: `char[] chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890".ToCharArray();` is good enough.

File: Okta.Samples.OpenIDConnect.Console/KeyGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            chars =
+            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890".ToCharArray();
+            byte[] data = new byte[1];
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.Console/KeyGenerator.cs]: `data` is initialized but later initialized with size of `maxSize`

File: Okta.Samples.OpenIDConnect.Console/KeyGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            using (RNGCryptoServiceProvider crypto = new RNGCryptoServiceProvider())
+            {
+                crypto.GetNonZeroBytes(data);
+                data = new byte[maxSize];
+                crypto.GetNonZeroBytes(data);
+            }
+            StringBuilder result = new StringBuilder(maxSize);
+            foreach (byte b in data)
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.Console/KeyGenerator.cs:23]: if removing the initializing on line 16 will give you a warning on 24, you could put the iteration within `using` block

File: Okta.Samples.OpenIDConnect.Console/Program.cs [hunk 4]
```diff
@@ -1,36 +1,73 @@
...
-                Console.WriteLine($"Result: {result}");
-                Console.WriteLine($"ID Token: {form.oAuthWebBrowser1.IDToken}");
-                Console.WriteLine($"Access Token: {form.oAuthWebBrowser1.AccessToken}");
-
-                Console.WriteLine("Application complete.");
-                Console.WriteLine("Press any key to exit...");
+                Console.WriteLine($"Result: {result}\r\n");
+                Console.WriteLine($"ID Token: {form.oAuthWebBrowser1.IDToken}\r\n");
+                Console.WriteLine($"Access Token: {form.oAuthWebBrowser1.AccessToken}\r\n");
+                foreach (Claim c in form.oAuthWebBrowser1.IDClaims)
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.Console/Program.cs:68]: missing `{}`

File: Okta.Samples.OpenIDConnect.Console/Program.cs [hunk 5]
```diff
@@ -40,5 +77,24 @@ static void Main(string[] args)
...
+                var discoAddress = "https://" + strOktaOrgDomain + "/.well-known/openid-configuration";
+
+                var manager = new ConfigurationManager<OpenIdConnectConfiguration>(discoAddress);
+
+                _config = await manager.GetConfigurationAsync();
+            }
+            return _config;
+        }
+
+        private static void ValidateToken(string jwtToken)
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.Console/Program.cs]: empty method?

File: Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        {
+            var discoAddress = strOktaOrgUrl;
+            if (bAuthorizationServer)
+            {
+                discoAddress = strOktaAuthZServerUrl;
+            }
+            discoAddress +=OIDC_CONFIG_SUFFIX;
+
+            OpenIdConnectConfiguration config = new OpenIdConnectConfiguration();
+
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs]: extra lines?

File: Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+        }
+
+        /// <summary>
+        /// Base64url no-padding encodes the given input buffer.
+        /// </summary>
+        /// <param name="buffer"></param>
+        /// <returns></returns>
+        public static string base64urlencodeNoPadding(byte[] buffer)
+        {
+            string base64 = Convert.ToBase64String(buffer);
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs]:
```
return Convert.ToBase64String(buffer)
            .Replace("+", "-")
            .Replace("/", "_")
            .Replace("=", "");
```

File: Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+                client.SetBearerToken(currentUser.FindFirst("access_token").Value);
+            }
+            string json = string.Empty;
+            try
+            {
+                json = await client.GetStringAsync(strApiEndpointUri);
+            }
+            catch (Exception ex)
+            {
+                string strEx = ex.ToString();
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs]: any usage for `strEx`?

File: Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+
+        private async Task<string> CallApi()
+        {
+            var client = new HttpClient();
+
+            if (currentUser.HasClaim(c => c.Type == "access_token"))
+            {
+                client.SetBearerToken(currentUser.FindFirst("access_token").Value);
+            }
+            string json = string.Empty;
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs]: nit: put 252 behind 246 and also use `var`

File: Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+        /// </summary>
+        /// <param name="code">the authorization code previously retrieved from the authorize end point</param>
+        /// <param name="code_verifier">the PKCE code previously generated during the authorize endpoint call</param>
+        async void performCodeExchange(string code, string code_verifier)
+        {
+            output("Exchanging code for tokens...");
+
+            // builds the token request
+            string tokenRequestURI = _oidcConfig.TokenEndpoint;
+            if(_authZConfig!=null)
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs]: space between `if` and `(`

File: Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+
+            TokenClient tokenClient = new TokenClient(tokenRequestURI, strClientId);
+
+            var tokenResponse = await tokenClient.RequestAuthorizationCodeAsync(code, strRedirectUri, code_verifier);
+
+            output("Making API Call to the UserInfo endpoint...");
+
+            // use the access token to retrieve claims from userinfo
+            string strUserInfoEndpoint = _oidcConfig.UserInfoEndpoint;
+            if(_authZConfig!=null)
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.NativeBrowserConsole/Program.cs]: space between `if` and `(`

File: Okta.Samples.OpenIDConnect.Console/Okta OpenID Connect Console (Implicit Flow - Embedded Browser).csproj [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    <DebugType>pdbonly</DebugType>
+    <Optimize>true</Optimize>
+    <OutputPath>bin\Release\</OutputPath>
+    <DefineConstants>TRACE</DefineConstants>
+    <ErrorReport>prompt</ErrorReport>
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <ItemGroup>
+    <Reference Include="IdentityModel, Version=2.1.1.0, Culture=neutral, processorArchitecture=MSIL">
+      <HintPath>..\packages\IdentityModel.2.1.1\lib\net45\IdentityModel.dll</HintPath>
```
Comment (yiyangli-okta) [inline Okta.Samples.OpenIDConnect.Console/Okta OpenID Connect Console (Implicit Flow - Embedded Browser).csproj:37]: did you forget to put nuget package config? I didn't see the dll file in the repo.

**Review by yiyangli-okta — CHANGES_REQUESTED** [review]

File: Okta.Samples.OpenIDConnect.Console/KeyGenerator.cs (hunk 2 above)
Comment (raphaellondner-okta) [reply Okta.Samples.OpenIDConnect.Console/KeyGenerator.cs]: Fixed with `byte[] data  = null;`

File: Okta.Samples.OpenIDConnect.Console/Program.cs (hunk 4 above)
Comment (raphaellondner-okta) [reply Okta.Samples.OpenIDConnect.Console/Program.cs:68]: Not sure what the issue is, I've checked this piece of code works

File: Okta.Samples.OpenIDConnect.Console/Program.cs (hunk 5 above)
Comment (raphaellondner-okta) [reply Okta.Samples.OpenIDConnect.Console/Program.cs]: removed

File: Okta.Samples.OpenIDConnect.Console/Program.cs (hunk 4 above)
Comment (yiyangli-okta) [reply Okta.Samples.OpenIDConnect.Console/Program.cs:68]: nothing wrong except the style, use a `{}` between and after an iteration body

File: Okta.Samples.OpenIDConnect.Console/Okta OpenID Connect Console (Implicit Flow - Embedded Browser).csproj (hunk 12 above)
Comment (raphaellondner-okta) [reply Okta.Samples.OpenIDConnect.Console/Okta OpenID Connect Console (Implicit Flow - Embedded Browser).csproj:37]: The NuGet package file is here --> https://github.com/raphaellondner-okta/okta-oauth-nativewindows/blob/google-console/Okta.Samples.OpenIDConnect.Console/packages.config

Comment (yiyangli-okta) [reply Okta.Samples.OpenIDConnect.Console/Okta OpenID Connect Console (Implicit Flow - Embedded Browser).csproj:37]: my bad, didn't see it.

File: Okta.Samples.OpenIDConnect.Console/Program.cs (hunk 4 above)
Comment (raphaellondner-okta) [reply Okta.Samples.OpenIDConnect.Console/Program.cs:68]: got it :)

**Comment by raphaellondner-okta** [comment]
@yiyangli-okta thanks for the quick review and feedback! Can take a peek at my latest [commit](https://github.com/raphaellondner-okta/okta-oauth-nativewindows/pull/2/commits/bca020962002742aa9ef317b4efd04c07ef1f164) based on your feedback?

**Comment by yiyangli-okta** [comment]
lgtm, please rebased 🚢

**Review by yiyangli-okta — APPROVED** [review]

**Pull request merged**